src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,757 @@
+                        otThreadErrorToString(error));
+        }
+
+        nextSolicitationDelay =
+            (mRouterSolicitCount == kMaxRtrSolicitations) ? kMaxRtrSolicitationDelay : kRtrSolicitationInterval;
+
+        otLogDebgBr("Router Solicitation timer scheduled in %u seconds", nextSolicitationDelay);
+        mRouterSolicitTimer.Start(nextSolicitationDelay * 1000);
+    }
+    else
+    {
+        // Re-evaluate our routing policy and send Router Advertisement if necessary.
+        EvaluateRoutingPolicy();
+    }
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<RoutingManager>().HandleDiscoveredOnLinkPrefixInvalidTimer();
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(void)
+{
+    otLogInfoBr("invalidate discovered on-link prefix: %s", mDiscoveredOnLinkPrefix.ToString().AsCString());
+    mDiscoveredOnLinkPrefix.Clear();
+
+    // The discovered on-link prefix becomes invalid, start Router Solicitation
+    // to discover new one.
+    StartRouterSolicitation();
+}
+
+void RoutingManager::HandleRouterSolicit(const Ip6::Address &aSrcAddress,
+                                         const uint8_t *     aBuffer,
+                                         uint16_t            aBufferLength)
+{
+    OT_UNUSED_VARIABLE(aSrcAddress);
+    OT_UNUSED_VARIABLE(aBuffer);
+    OT_UNUSED_VARIABLE(aBufferLength);
+
+    otLogInfoBr("received Router Solicitation from %s on interface %s", aSrcAddress.ToString().AsCString(),
+                GetInfraIfName());
+
+    mRouterAdvertisementTimer.Start(Random::NonCrypto::GetUint32InRange(0, kMaxRaDelayTime));
+}
+
+uint32_t RoutingManager::GetPrefixExpireDelay(uint32_t aValidLifetime)
+{
+    uint32_t delay;
+
+    if (aValidLifetime * static_cast<uint64_t>(1000) > Timer::kMaxDelay)
+    {
+        delay = Timer::kMaxDelay;
+    }
+    else
+    {
+        delay = aValidLifetime * 1000;
+    }
+
+    return delay;
+}
+
+void RoutingManager::HandleRouterAdvertisement(const Ip6::Address &aSrcAddress,
+                                               const uint8_t *     aBuffer,
+                                               uint16_t            aBufferLength)
+{
+    OT_UNUSED_VARIABLE(aSrcAddress);
+
+    using RouterAdv::Option;
+    using RouterAdv::PrefixInfoOption;
+    using RouterAdv::RouterAdvMessage;
+
+    bool           needReevaluate = false;
+    const uint8_t *optionsBegin;
+    uint16_t       optionsLength;
+    const Option * option;
+
+    VerifyOrExit(aBufferLength >= sizeof(RouterAdvMessage));
+
+    otLogInfoBr("received Router Advertisement from %s on interface %s", aSrcAddress.ToString().AsCString(),
+                GetInfraIfName());
+
+    optionsBegin  = aBuffer + sizeof(RouterAdvMessage);
+    optionsLength = aBufferLength - sizeof(RouterAdvMessage);
+
+    option = Option::GetNextOption(nullptr, optionsBegin, optionsLength);
+    while (option != nullptr)
```

Comment:
Suggestion: How about combining this into one `while()` command, with this we can remove the call to `GetNextOption()` at the end of loop.

```suggestion
    option = nullptr;
    while ((option = Option::GetNextOption(nullptr, optionsBegin, optionsLength)) != nullptr)
```

The other thing we can do with this change, is to use `continue` within the block to skip over when some condition we expect is not met to go to next option. e.g. we can consider adding a `if (!IsValidOnLinkPrefix(prefix))` then add the log and do `continue` instead of doing it from `else`.